temporalio/worker/_workflow.py

```diff
@@ -250,9 +305,31 @@ async def _handle_activation(
                         activate_task, self._deadlock_timeout_seconds
                     )
                 except asyncio.TimeoutError:
-                    raise RuntimeError(
-                        f"[TMPRL1101] Potential deadlock detected, workflow didn't yield within {self._deadlock_timeout_seconds} second(s)"
-                    )
+                    # extract the stack of the workflow if possible
+
+                    # cases:
+                    # 1. get_thread_id() is not implemented for the worker (returns None, fall to case 2)
+                    # 2. get_thread_id() returns None
+                    # 3. tid is not present in sys._current_frames()
+                    # 4. error occurs during traceback construction
+                    # 5. normal
+
+                    msg = f"[TMPRL1101] Potential deadlock detected, workflow didn't yield within {self._deadlock_timeout_seconds} second(s)."
+                    tid = workflow.get_thread_id()
+                    if not tid:
+                        # Default error if the function is not implemented for the workers, or the worker is not running
+                        raise RuntimeError(msg) from None
+
+                    tb, warn = get_traceback_for_thread(tid)
+
+                    if not tb:
+                        msg += "Worker was unable to retrieve workflow thread info. No frames from workflow available."
+                        raise RuntimeError(msg) from None
+                    else:
+                        if tb and warn:
+                            msg += "System encountered errors while constructing traceback, minimal or partial traceback will be presented."
+                        raise RuntimeError(msg).with_traceback(tb) from None
+
```

Comment:
nit: it's slightly better style to handle the non-negated case first (or handle negated case first and then don't bother with the `else` to reduce nesting, as one would in Go).